### PR TITLE
add node-gpu controller to get GPU info from node and pod

### PR DESCRIPTION
**Pull request by nowinkeyy**
### Summary
🐳🛠️📝

This pull request adds the initial scaffolding and configuration for the node controller project, which is a Kubernetes controller that manages node resources. The pull request uses kubebuilder and kustomize to generate the code, the manifests, the Dockerfile, and the Makefile for the controller. The pull request also adds some RBAC and auth proxy features, as well as a prometheus monitor service, for the controller. The pull request also includes a `.gitignore` and a `.dockerignore` file, as well as a `README.md` file with some basic documentation.

> _`node` controller_
> _docker, kustomize, rbac_
> _autumn scaffolding_

### Walkthrough
*  Add a node controller project with kubebuilder scaffolding and configuration ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-a1fabca8c36ab2c4df7e970912782f7d4d08c6bade6cbdb3ac6b8954fe918f3aR1-R13), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-56f424305b3939716c0eb3cd817c96239a5401e80e07e042dc501f7d0737825bR1-R94))
*  Define a Dockerfile to build a distroless image for the node controller ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-2729d1f7a8f277629af48ceaa75b07bd0db65b42e84b7a9c10bc97ed4aba81f8R1-R4), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-86a31f6242bc842b5e44fbcb523dfbc04b2ad1133031dd3f379118e9d8665c78L1-R7))
*  Define a Makefile to provide various targets for building, testing, generating, and deploying the node controller ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-1074b22ab980cce71ed99005c4b7c1c4b027874bbaa76a78a24cd4a579e75513R1-R26), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-db3b1c8920c8f983a9eeda9cf08de2368d66633e1d0eb55034757f62df6e348eR1-R139))
*  Define a kustomization file and a deployment manifest for the manager container ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-fb6e0c57a6294c2822809e7f23dcbb510a862115bf1de3ce78f0d355d067b7b0R1-R8), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-bf0ea9fe083057d0c42cdd308c8104696d9d5a0e8f5ea3b6e75d7185a96cfe0eR1-R102))
*  Define a kustomization file and a patch to inject an auth proxy sidecar container for RBAC authorization ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-ba22a504df5e486758c211d5bb5389f347a5cb55d079eb527a2c5dc75eb975d5R1-R144), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-ce3f5a73ff92f201dde6952b2e4c6b0012e375469402184bd93c00ce3ef95e58R1-R55))
*  Define a kustomization file and a patch to configure the manager container ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-ba22a504df5e486758c211d5bb5389f347a5cb55d079eb527a2c5dc75eb975d5R1-R144), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-01a349d66c95e56f03dcd4e1c3579b346d94931b6917a4fbf8b885973ab84a8bR1-R10))
*  Define a kustomization file and a service monitor resource for the prometheus operator ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-929c7d29b557a237f6e46b8d2cd3d7806e5cc3b00377312cf02de445b9e2e3e4R1-R2), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-70965fce87647d77c37fe202c362f0dbfa87cf33e5fab85a6ddc9064f7c81428R1-R26))
*  Define a kustomization file and various RBAC resources for the controller manager and the auth proxy ([link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-2c1a61dddd30abdadb5d534292308039f97b0d15ec7f820f2aa80f7ec56f9bf4R1-R18), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-c3f0d1df8f3ffeb3cd13d318f48e7bf2b8ce98c66424e808d2f5d7b345f09805R1-R16), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-b13124a95e580f567fda72543a23c467f2f8ee758efe75f330d31374f0c0b595R1-R24), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-5eae0dcef034c4fe220d2f4dd6bdda1adf96acbc1226fde05e26b0d9e23589ddR1-R19), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-38632000fd8be064eaff503c74da235391db1535dfdc8ad8bf76b05866f73f3eR1-R21), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-0117bc6a490c10eb83bf9d770f5702ac0cff29b0355708f988f9d20eaccc8157R1-R44), [link](https://github.com/labring/sealos/pull/3601/files?diff=unified&w=0#diff-2db5af1c13da689ccba2c7f6e4887ca4ad5d766bbf0c0e92be829049bbaeecaeR1-R19))

### CODE REVIEW

**Comment by dinoallo** [comment]
LGTM. Since `node` is ambiguous, maybe it's better to rename the controller to `node-gpu` or `gpu-counter` to avoid confusion?

File: controllers/node/controllers/gpu_controller.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+func (r *GpuReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	nodeList := &corev1.NodeList{}
+	if err := r.Client.List(ctx, nodeList); err != nil {
+		r.Logger.Error(err, "failed to get node list")
+		return ctrl.Result{}, err
+	}
+
+	// todo index pods on GPU node
+	podList := &corev1.PodList{}
```
Comment (zzjin) [inline controllers/node/controllers/gpu_controller.go:70]: get all pods in every reconcile seem too heavy?

File: controllers/node/controllers/gpu_controller.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+//+kubebuilder:rbac:groups=node.k8s.io,resources=gpus,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=node.k8s.io,resources=gpus/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=node.k8s.io,resources=gpus/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=nodes/status,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+func (r *GpuReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	nodeList := &corev1.NodeList{}
```
Comment (zzjin) [inline controllers/node/controllers/gpu_controller.go:63]: The node is not get updated frequently, so add cache+informer will almost prevent all calls to `api-server`.

**Comment by nowinkeyy** [comment]
Test the performance difference before and after index creation.
(elapsed1/elapsed2 is the time spent after optimization, elapsed11/elapsed22 is the time spent before optimization)

- index type: node
- index/total = 1/5
image
image

- index type: pod
- index/total = 1/10
image

- index/total = 1/100
image

**Review by zzjin — APPROVED** [review]

**Comment by bxy4543** [comment]
It also needs to be built as a sealos cluster-image and added to workflow automation . @nowinkeyy

**Pull request merged**